// payload/dist/globals/graphql/resolvers/update.js
/* eslint-disable no-param-reassign */ "use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return updateResolver;
    }
});
const _isolateObjectProperty = /*#__PURE__*/ _interop_require_default(require("../../../utilities/isolateObjectProperty"));
const _update = /*#__PURE__*/ _interop_require_default(require("../../operations/update"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function updateResolver(globalConfig) {
    return async function resolver(_, args, context) {
        if (args.locale) context.req.locale = args.locale;
        if (args.fallbackLocale) context.req.fallbackLocale = args.fallbackLocale;
        const { slug } = globalConfig;
        const options = {
            data: args.data,
            depth: 0,
            draft: args.draft,
            globalConfig,
            req: (0, _isolateObjectProperty.default)(context.req, 'transactionID'),
            slug
        };
        const result = await (0, _update.default)(options);
        return result;
    };
}
